### PR TITLE
i#1870: Support ordinal forwarders

### PR DESCRIPTION
**Pull request by woodruffw**
Adds support for ordinal forwarders, and fixes a sanity check to
handle DLLs that contain no named exports.

In progress, but I might not have much time to work on this in the near future. Someone else who runs into this issue should feel free to adopt these changes if they need them immediately.

### CODE REVIEW

**Comment by AssadHashmi** [comment]
run arm tests

**Comment by woodruffw** [comment]
This should also be ready for review 😄

**Comment by derekbruening** [comment]
Generally, do not force-push to a shared branch (our wiki states this as well): it ruins the history, messes up anyone who has checked out this branch, and loses review history.  Once shared, append only.

File: core/win32/module_shared.c [hunk 1]
```diff
@@ -371,17 +371,35 @@ get_proc_address_common(module_base_t lib, const char *name,
 
     exports = get_module_exports_directory_check_common(module_base, &exports_size,
                                                         true _IF_NOT_X64(ldr64));
-    if (exports == NULL || exports_size == 0 || exports->NumberOfNames == 0 ||
-        /* just extra sanity check */ exports->AddressOfNames == 0)
+
+    /* NB: There are some DLLs (like System32\profapi.dll) that have no named
+     * exported functions names, only ordinals. As a result, the only correct
+     * checks we can do here are on the presence and size of the export table
+     * + presenence and count of the function export list.
```
Comment (derekbruening) [inline core/win32/module_shared.c]: spelling: s/+ presenence/and the presence/

File: core/win32/module_shared.c [hunk 2]
```diff
@@ -371,17 +371,35 @@ get_proc_address_common(module_base_t lib, const char *name,
...
 #if !defined(NOT_DYNAMORIO_CORE_PROPER) && !defined(NOT_DYNAMORIO_CORE)
-    /* sanity check */
-    ASSERT(exports->AddressOfFunctions < module_size && exports->AddressOfFunctions > 0 &&
-           exports->AddressOfNameOrdinals < module_size &&
-           exports->AddressOfNameOrdinals > 0 && exports->AddressOfNames < module_size &&
-           exports->AddressOfNames > 0);
+    /* sanity checks, split up for readability */
+
+    /* The DLL either exports nothing or has a sane combination of export
+     *  table address and function count.
```
Comment (derekbruening) [inline core/win32/module_shared.c]: nit: extra space

File: core/win32/module_shared.c [hunk 3]
```diff
@@ -393,6 +411,31 @@ get_proc_address_common(module_base_t lib, const char *name,
...
+                 */
+                ASSERT_CURIOSITY(false && "non-numeric ordinal forwarder");
+                return NULL;
+            }
+
+            ord *= 10;
+            ord += (digit[0] - '0');
+
+            digit += 1;
+        }
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply core/win32/module_shared.c]: Sounds good to me -- perhaps `%u`, since ordinals should never be negative as well?

**Comment by woodruffw** [comment]
Realized that I didn't address the space nit. This should be ready for another look, thanks!

File: core/win32/module_shared.c [hunk 4]
```diff
@@ -375,7 +377,7 @@ get_proc_address_common(module_base_t lib, const char *name,
     /* NB: There are some DLLs (like System32\profapi.dll) that have no named
      * exported functions names, only ordinals. As a result, the only correct
      * checks we can do here are on the presence and size of the export table
-     * + presenence and count of the function export list.
+     * abd the presence and count of the function export list.
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply core/win32/module_shared.c]: Fixed, thanks.

**Comment by derekbruening** [comment]
run arm tests

**Comment by woodruffw** [comment]
Thanks!

**Pull request merged**